src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/charts/utils.ts

```diff
@@ -54,3 +55,21 @@ export function getDatasourceLayers(
     ...(indexPatternLayers && Object.keys(indexPatternLayers).length ? indexPatternLayers : {}),
   };
 }
+
+export function getLensStateLayer(
+  layers: Record<string, Omit<FormBasedLayer, 'indexPatternId'> | TextBasedLayer>,
+  visLayerId: string | undefined
+) {
+  // Filter to keep non-linked layers (layers without linkToLayers or with linkToLayers set to null)
+  // Also keep ES|QL layers with non-empty columns, as old layers persist after chart type switches and have empty columns
+  const mainLayers = Object.entries(layers).filter(
+    ([, l]) =>
+      !('linkToLayers' in l) ||
+      l.linkToLayers == null ||
+      (isTextBasedLayer(l) && 'columns' in l && Array.isArray(l.columns) && l.columns.length > 0)
```

Comment:
nit: given the `isTextBasedLayer` check, can we drop the middle checks?


```suggestion
      (isTextBasedLayer(l) && l.columns.length > 0)
```